crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -509,37 +551,52 @@ impl PgManager {
             .map_err(|e| Error::Internal(e.to_string()))
     }
 
-    /// Takes a query fragment and a lambda that executes the query
+    /// Takes a query_builder_fn that returns Result<QueryFragment> and a lambda to execute the query
     /// Spawns a blocking task that determines the cost of the query fragment
-    /// And if within limits, executes the query
-    async fn run_query_async_with_cost<T, Q, EF, E, F>(
+    /// And if within limits, then executes the query
+    async fn run_query_async_with_cost<T, Q, QResult, EF, E, F>(
         &self,
-        query: Q,
+        mut query_builder_fn: Q,
         execute_fn: EF,
     ) -> Result<T, Error>
     where
-        Q: QueryDsl
+        Q: FnMut() -> Result<QResult, Error> + Send + 'static,
+        QResult: QueryDsl
             + RunQueryDsl<Pg>
             + diesel::query_builder::QueryFragment<diesel::pg::Pg>
+            + diesel::query_builder::Query
+            + diesel::query_builder::QueryId
             + Send
             + 'static,
-        EF: FnOnce(Q) -> F + Send + 'static,
+        EF: FnOnce(QResult) -> F + Send + 'static,
         F: FnOnce(&mut PgConnection) -> Result<T, E> + Send + 'static,
         E: From<diesel::result::Error> + std::error::Error + Send + 'static,
         T: Send + 'static,
     {
         let max_db_query_cost = self.limits.max_db_query_cost;
         self.inner
             .spawn_blocking(move |this| {
-                let cost = extract_cost(&query, &this)?;
-                if cost > max_db_query_cost as f64 {
-                    return Err(DbValidationError::QueryCostExceeded(
-                        cost as u64,
-                        max_db_query_cost,
-                    )
-                    .into());
+                let query = query_builder_fn()?;
+                let explain_result: String = this
+                    .run_query(|conn| query.explain().get_result(conn))
+                    .map_err(|e| Error::Internal(e.to_string()))?;
+                let parsed: serde_json::Value = serde_json::from_str(&explain_result)
+                    .map_err(|e| Error::Internal(e.to_string()))?;
+                if let Some(cost) = parsed[0]["Plan"]["Total Cost"].as_f64() {
```

Comment:
Do we need to be careful about how we access the cost?  Indexing in Rust usually panics if the key does not exist, is there a way we can chain together `get`s or something?

Reply:
Let me look into that

Reply:
I've split out the parsing into its own function and added some tests, thanks for calling this out